src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -0,0 +1,430 @@
+package io.iohk.ethereum.blockchain.sync.regular
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props, Scheduler}
+import akka.util.ByteString
+import cats.Eq
+import cats.syntax.either._
+import cats.syntax.eq._
+import io.iohk.ethereum.blockchain.sync.PeerRequestHandler.{RequestFailed, ResponseReceived}
+import io.iohk.ethereum.blockchain.sync.regular.BlockImporter.{ImportNewBlock, ImporterMsg, NotOnTop, OnTop}
+import io.iohk.ethereum.blockchain.sync.{BlacklistSupport, PeerListSupport, PeerRequestHandler}
+import io.iohk.ethereum.crypto.kec256
+import io.iohk.ethereum.db.storage.TotalDifficultyStorage.TotalDifficulty
+import io.iohk.ethereum.domain.{Block, BlockHeader}
+import io.iohk.ethereum.network.EtcPeerManagerActor.PeerInfo
+import io.iohk.ethereum.network.Peer
+import io.iohk.ethereum.network.PeerEventBusActor.PeerEvent.MessageFromPeer
+import io.iohk.ethereum.network.PeerEventBusActor.SubscriptionClassifier.MessageClassifier
+import io.iohk.ethereum.network.PeerEventBusActor.{PeerSelector, Subscribe}
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.NewBlock
+import io.iohk.ethereum.network.p2p.messages.PV62._
+import io.iohk.ethereum.network.p2p.messages.PV63.{GetNodeData, NodeData}
+import io.iohk.ethereum.network.p2p.{Message, MessageSerializable}
+import io.iohk.ethereum.utils.Config.SyncConfig
+import mouse.all._
+
+import scala.collection.immutable.Queue
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+
+class BlockFetcher(
+    val etcPeerManager: ActorRef,
+    val peerEventBus: ActorRef,
+    val syncConfig: SyncConfig,
+    implicit val scheduler: Scheduler
+) extends Actor
+    with ActorLogging
+    with PeerListSupport
+    with BlacklistSupport {
+
+  import BlockFetcher._
+
+  val bufferSize = 10000
+
+  scheduler.schedule(1.second, 1.second, self, PrintStatus)
+  peerEventBus ! Subscribe(MessageClassifier(Set(NewBlock.code, NewBlockHashes.code), PeerSelector.AllPeers))
+
+  override def receive: Receive = idle(FetcherState.initial)
+
+  private def idle(state: FetcherState): Receive =
+    handleCommonMessages(state, idle) orElse {
+      case Start(blockNr) => state |> FetcherState.withLastBlock(blockNr) |> fetchBlocks
```

Comment:
It's usually called "pipeline operator". It's defined in mouse - bunch of helpers to cats. It works like pipe in bash: takes value on the left and apply it to function on the right side. It's really cool in situations where couple of functions needs to be composed ad-hoc to compute some result. It also allows to avoid temporary variables for intermediate steps (which can be very hard to name properly)